core/tests/ts-integration/src/retry-provider.ts

```diff
@@ -20,100 +20,192 @@ function isIgnored(err: any): boolean {
+            `Created L2 transaction ${base.hash} (from=${base.from}, to=${base.to}, ` +
+                `nonce=${base.nonce}, value=${base.value}, data=${debugData(base.data)})`
+        );
+    }
+    knownTransactionHashes.add(base.hash);
+}
+
+function _handleTransactionReceipt(knownTransactionHashes: Set<string>, reporter: Reporter, receipt: any) {
+    if (!knownTransactionHashes.has(receipt.transactionHash)) {
+        knownTransactionHashes.add(receipt.transactionHash);
+        reporter.debug(
+            `Obtained receipt for L2 transaction ${receipt.transactionHash}: blockNumber=${receipt.blockNumber}, status=${receipt.status}`
+        );
+    }
+}
+
+/**
+ * RetryProvider retries every RPC request if it detects a timeout-related issue on the server side.
+ */
+export class RetryProvider extends zksync.Provider {
+    readonly reporter: Reporter;
+    private readonly knownTransactionHashes: Set<string> = new Set();
+
+    constructor(url: string | { url: string; timeout: number }, network?: ethers.Networkish, reporter?: Reporter) {
+        const fetchRequest = createFetchRequest(url, () => this.pollingInterval, reporter);
+        super(fetchRequest, network);
+        this.reporter = reporter ?? new Reporter();
+    }
+
+    override async send(method: string, params: any): Promise<any> {
+        return _send(
+            (method, params) => super.send(method, params),
+            this.pollingInterval,
+            this.reporter,
+            method,
+            params
+        );
     }
 
     override _wrapTransactionResponse(txResponse: any): L2TransactionResponse {
         const base = super._wrapTransactionResponse(txResponse);
-        const isNew = !this.knownTransactionHashes.has(base.hash);
-        if (isNew) {
-            this.reporter.debug(
-                `Created L2 transaction ${base.hash} (from=${base.from}, to=${base.to}, ` +
-                    `nonce=${base.nonce}, value=${base.value}, data=${debugData(base.data)})`
-            );
-        }
-
-        this.knownTransactionHashes.add(base.hash);
+        _handleTransactionResponse(this.knownTransactionHashes, this.reporter, base);
         return new L2TransactionResponse(base, this.reporter);
     }
 
     override _wrapTransactionReceipt(receipt: any): zksync.types.TransactionReceipt {
-        const wrapped = super._wrapTransactionReceipt(receipt);
-        if (!this.knownTransactionHashes.has(receipt.transactionHash)) {
-            this.knownTransactionHashes.add(receipt.transactionHash);
-            this.reporter.debug(
-                `Obtained receipt for L2 transaction ${receipt.transactionHash}: blockNumber=${receipt.blockNumber}, status=${receipt.status}`
-            );
-        }
-        return wrapped;
+        _handleTransactionReceipt(this.knownTransactionHashes, this.reporter, receipt);
+        return super._wrapTransactionReceipt(receipt);
+    }
+}
+
+export class EthersRetryProvider extends ethers.JsonRpcProvider {
```

Comment:
Thanks for not forgetting about adapting `RetryProvider` stuff :)

Reply:
It was actually present before in another module, but I've largely forgotten about it 🙃 Should be fixed now.